hazelcast/src/test/java/com/hazelcast/map/impl/nearcache/invalidation/MemberMapReconciliationTest.java

```diff
@@ -0,0 +1,115 @@
+import static java.lang.String.valueOf;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class MemberMapReconciliationTest extends HazelcastTestSupport {
+
+    private static final String MAP_NAME = "test";
+    private static final int RECONCILIATION_INTERVAL_SECONDS = 3;
+
+    private final TestHazelcastInstanceFactory factory = new TestHazelcastInstanceFactory();
+    private final Config config = new Config();
+
+    private IMap serverMap;
+    private IMap nearCachedServerMap;
+
+    @Before
+    public void setUp() throws Exception {
+        NearCacheConfig nearCacheConfig = new NearCacheConfig(MAP_NAME);
+        nearCacheConfig.setInvalidateOnChange(true);
+        nearCacheConfig.setCacheLocalEntries(true);
+
+        config.setProperty("hazelcast.invalidation.max.tolerated.miss.count", "0");
+        config.setProperty("hazelcast.invalidation.reconciliation.interval.seconds", valueOf(RECONCILIATION_INTERVAL_SECONDS));
+        config.setProperty("hazelcast.invalidation.min.reconciliation.interval.seconds", valueOf(RECONCILIATION_INTERVAL_SECONDS));
+        config.setProperty(MAP_INVALIDATION_MESSAGE_BATCH_FREQUENCY_SECONDS.getName(), valueOf(Integer.MAX_VALUE));
+        config.setProperty(MAP_INVALIDATION_MESSAGE_BATCH_SIZE.getName(), valueOf(Integer.MAX_VALUE));
+        config.getMapConfig(MAP_NAME).setNearCacheConfig(nearCacheConfig);
+
+        HazelcastInstance server = factory.newHazelcastInstance(config);
+        serverMap = server.getMap(MAP_NAME);
+
+        nearCachedServerMap = nearCachedMapFromNewServer();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        factory.shutdownAll();
+    }
+
+    private IMap nearCachedMapFromNewServer() {
+        HazelcastInstance server = factory.newHazelcastInstance(config);
+        IMap map = server.getMap(MAP_NAME);
+
+        assert map instanceof NearCachedMapProxyImpl;
+
+        return map;
+    }
+
+    @Test
+    public void test_reconciliation_does_not_cause_premature_removal() throws Exception {
+        int total = 100;
+        for (int i = 0; i < total; i++) {
+            serverMap.put(i, i);
+        }
+
+        for (int i = 0; i < total; i++) {
+            nearCachedServerMap.get(i);
+        }
+
+        IMap nearCachedMapFromNewServer = nearCachedMapFromNewServer();
+        for (int i = 0; i < total; i++) {
+            nearCachedMapFromNewServer.get(i);
+        }
+
+        NearCacheStats nearCacheStats = nearCachedMapFromNewServer.getLocalMapStats().getNearCacheStats();
+
+        assertEquals(total, nearCacheStats.getOwnedEntryCount());
+
+        sleepSeconds(2 * RECONCILIATION_INTERVAL_SECONDS);
+
+        for (int i = 0; i < total; i++) {
+            nearCachedMapFromNewServer.get(i);
+        }
+
+        assertEquals(total, nearCacheStats.getMisses());
```

Comment:
Same question: can we add owned entry count and hits checks here? Same for the other test variants.